java-extras/src/test/java/org/triplea/java/cache/ExpiringAfterWriteCacheTest.java

```diff
@@ -0,0 +1,204 @@
+package org.triplea.java.cache;
+
+import static com.github.npathai.hamcrestopt.OptionalMatchers.isEmpty;
+import static com.github.npathai.hamcrestopt.OptionalMatchers.isPresentAndIs;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.Is.is;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.atLeastOnce;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.triplea.java.cache.TtlCache.CacheEntry;
+
+@ExtendWith(MockitoExtension.class)
+class ExpiringAfterWriteCacheTest {
+
+  private static final String KEY = "key-value";
+  private static final int VALUE = 100;
+  private ExpiringAfterWriteCache<String, Integer> realCache;
+
+  @Mock private Consumer<CacheEntry<String, Integer>> cacheRemovalListener;
+
+  @BeforeEach
+  void setup() {
+    realCache = new ExpiringAfterWriteCache<>(1, TimeUnit.MINUTES, cacheRemovalListener);
+  }
+
+  @Nested
+  class GetAndPut {
+    @Test
+    void getWillReturnValuesThatHaveBeenPutIntoCache() {
+      realCache.put("id", 0);
+      assertThat(realCache.get("id"), isPresentAndIs(0));
+    }
+
+    @Test
+    void getReturnsEmptyIfValueDoesNotExist() {
+      assertThat(realCache.get("DNE"), isEmpty());
+    }
+
+    @Test
+    void putCanBeUsedToReplace() {
+      realCache.put("id", 0);
+
+      realCache.put("id", 1);
+
+      assertThat(realCache.get("id"), isPresentAndIs(1));
+      verify(cacheRemovalListener, never()).accept(any());
```

Comment:
Thinking about my suggestion that passes a custom executor for the tests: This particular tests would gain some extra safety, because in case the removalListener is executed asynchronously, the tests might check too early missing this critical error.

Reply:
Agree the test could miss the timing. In practice this won't be a problem, there is a trade-off here between a slow test and more thorough test. Adding a second delay would be the thing to do, and/or making the cache scheduler really aggressive.

I double checked with a timeout and many repeated tests, as expected the test stays solid which is the most important thing. At this point I'm happy leaving the test to be more behavior documentation/spec than full-proof.

Reply:
Another approach than specifying a direct executor is to use a testing library like [Awaitility](https://github.com/awaitility/awaitility/wiki/Usage).

Reply:
Really what we are testing is the if condition here: https://github.com/triplea-game/triplea/pull/6007/files#diff-bd3a571488048aa02a100236fc5d8a77R33

```
            .removalListener(
                (IdT key, ValueT value, RemovalCause cause) -> {
                  if (cause == RemovalCause.EXPIRED || cause == RemovalCause.EXPLICIT) {
                    removalListener.accept(key, value);
                  }
```

Mockito 'verify' has a timeout function, so it would be straightforward to add the delay. The issue is that we are testing the absence of a call, so we'd always hit the timeout making this a very slow test.

IMO the best way to test probably would be to inject the filter logic, test the filter, and then trust that the library code does call the filter on remove listener.

I'm leaning towards keeping this test as is:
- injecting modules is at danger of over-engineering
- the ROI of a slow but more thorough test is not favorable. We're already close to testing pure library code